Add support for MINETEST_USERDATA environment variable

vilhelmgray opened the pull request:
The MINETEST_USERDATA environment variable can be used to defined a
custom path for Minetest user data. If MINETEST_USERDATA is empty or
unset, the HOME environment variable is used as the default user data
path; this ensures backwards compatibility with existing user setups.

SmallJoker commented:
Please document this feature here: https://github.com/minetest/minetest/blob/master/doc/minetest.6

Also a notice in "Paths" (https://github.com/minetest/minetest/blob/master/README.md#paths) might be helpful - either a general one, or mentioning the command line and environment variable methods.

vilhelmgray commented:
I decided to change the variable name to `MINETEST_USER_PATH` to match the naming convention of the existing Minetest environment variables listed in the [`minetest.6`](https://github.com/minetest/minetest/blob/master/doc/minetest.6) file. I also added support for an equivalent `%MINETEST_USER_PATH%` Windows environment variable so users on Windows can also set this path.

SmallJoker reviewed (approved):
LGTM

Df458 reviewed (changes requested):
I find it a little odd that `$MINETEST_USER_PATH` is not actually the path to minetest's user directory, but rather the path to _the directory where minetest's user directory is created_. My first reaction upon seeing this was "oh cool there's now a way to put my tests the data in `.local/share`" but you can only _kinda_ do that... because it'll still make a hidden directory. For that reason, my first instinct when testing was to try `~/.local/share/minetest`... but in fact this will fail, because MT will only try to create the .minetest dir and not any of its parents. This yields the not very useful error of
> Cannot create user data directory

I feel like if we're going to have a feature like this, we should
- Allow minetest to create the full path, not just the final directory
- Let the user specify the full path to the data, not the path to `.minetest`'s parent.

If we don't do this, I think the documentation should be updated. "Path to Minetest user data directory." doesn't read as the parent dir, whereas something like "Path where the Minetest user data directory will be created" might be clearer. I will leave that determination to the PR author and/or any other coredevs who want to chime in.

vilhelmgray commented:
I agree that specifying the actual directory would be better. The reason for the current behavior of Minetest creating a `.minetest` folder is because the user's home directory is assumed and thus we don't want to pollute the user's home directory with various Minetest data. If the user can now specify the directory they want, there's no need to create a `.minetest` directory under that because the user has already chosen where they want Minetest data to go.

However before I update this, I'll wait for other devs to chime in and confirm this is a good approach to take for this PR.

SmallJoker commented:
I totally missed that part about nested directories. Indeed, I agree with @Df458 that the variable should point directly to the user data, and not its parent directory.

vilhelmgray commented:
So I adjusted the code to have `path_user` set to the `MINETEST_USER_DATA` environment variable directory (if it is set).

In separate commit, I adjusted [`create_userdata_path()`](https://github.com/minetest/minetest/blob/0e439b2fa3f0ba4f3352273519c32ebc9c3a680e/src/main.cpp#L501) to create the user data directory using [`std::filesystem::create_directories()`](https://en.cppreference.com/w/cpp/filesystem/create_directory); this will automatically create any missing parent directories. The problem is that this function was added in **C++17**, but I think the Minetest [CMakeLists.txt](https://github.com/minetest/minetest/blob/0e439b2fa3f0ba4f3352273519c32ebc9c3a680e/CMakeLists.txt#L14) specifies **C++14**. Does it make sense to bump up the C++ standard for Minetest, or does another part of the codebase prevent that?

SmallJoker commented:
You could recursively call `fs::RemoveLastPathComponent` with `fs::IsDir` to determine which directories to create. The switch to C++14 was done recently, and it will again take a while until most of the supported distributions have GCC with C++17 support.

vilhelmgray commented:
Okay, I understand. I'll utilize `fs::RemoveLastPathComponent` instead then.

vilhelmgray commented:
As I was implementing this, I discovered `fs::CreateAllDirs` which does what we want so I've adjusted the commit to call that instead.

SmallJoker reviewed (approved):
Tested on Linux with a trivial change in `initializePaths()` to enforce system paths due to RUN_IN_PLACE configuration/build.

Works. The environment variable now points exactly to the desired location.

sfan5 commented on README.md:
```diff
@@ -95,18 +95,18 @@ Where each location is on each platform:
     * `share` = `/usr/share/minetest`
-    * `user`  = `~/.minetest`
-* macOS:
+    * `user`  = `$MINETEST_USER_PATH`
```
It'd be much clearer to put the fallback value first
```suggestion
    * `user`  = `~/.minetest` or `$MINETEST_USER_PATH`
```

sfan5 commented on src/porting.cpp:
```diff
@@ -422,11 +422,16 @@ bool setSystemPaths()
-	DWORD len = GetEnvironmentVariable("APPDATA", buf, sizeof(buf));
-	FATAL_ERROR_IF(len == 0 || len > sizeof(buf), "Failed to get APPDATA");
+	// Use %MINETEST_USER_PATH%
+	DWORD len = GetEnvironmentVariable("MINETEST_USER_PATH", path_user.data(), path_user.size());
```
path_user is empty when this is called so this will never work correctly

sfan5 commented on src/porting.cpp:
```diff
@@ -486,8 +491,11 @@ bool setSystemPaths()
 #ifndef __ANDROID__
-	path_user = std::string(getHomeOrFail()) + DIR_DELIM "."
-		+ PROJECT_NAME;
+	path_user = std::string(getenv("MINETEST_USER_PATH"));
```
getenv returns null if unset, constructing a string from a null pointer is undefined behaviour

sfan5 reviewed (changes requested):
If you look at CI you will also see that the build is broken

sfan5 reviewed (approved):
Looks good.
